EIPS/eip-7814.md

```diff
@@ -0,0 +1,59 @@
+---
+eip: 7814
+title: Introspection precompiles
+description: Introspection precompiles that expose the current block context to the EVM
+author: Brecht Devos (@Brechtpd)
+discussions-to: https://ethereum-magicians.org/t/eip-7814-introspection-precompiles/21872
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-11-09
+---
+
+## Abstract
+
+This EIP proposes to add two precompiles that enable introspection of the chain state at arbitrary points within a block in the EVM. Currently, the EVM only has access to the state of previous blocks. No block data is currently exposed to the EVM for the block it's executing in.
+
+## Motivation
+
+The new precompiles aim to enhance introspection capabilities within the EVM, enabling the calculation of the latest chain state offchain at any point in an Ethereum block. This is important to allow general and efficient synchronous composability with L1. Otherwise, to ensure having the latest L1 state, the state would have to be read on L1 and passed in as a separate input. This is expensive and there may be limitations on who can read the state without something like [EIP-2330](https://eips.ethereum.org/EIPS/eip-2330).
+
+This proposal allows computing the latest state from the state root of the previous block and the transactions that are in the current block. This data can then be passed into any system requiring the latest chain state where the partial block can be re-executed to compute the latest state in a provable way.
+
+## Specification
+
+If `block.number >= TBD` two new precompiled contracts shall be created:
+- `TXTRIEROOT` at address `TBD`: This precompile returns the transaction trie root of all transactions in the current block, up to and including the transaction that is currently executing. The tx trie is already calculated for the block header. This EIP just enforces that the trie is constructed incrementally per transaction and exposes the root to the EVM.
```

Comment:
seems that it would be technically feasible to only compute stateroot till previous tx. kinky confirm that what you wrote is what you intend to propose

Reply:
Changed behavior and clarified (hopefully)!